### PR TITLE
Add RobustSoliton distribution and benchmarking capabilities

### PR DESCRIPTION
**Pull request by vihu**

### CODE REVIEW

File: benches/decoder_benchmark.rs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    let sizes: Vec<usize> = (1000..1002).collect();
+    let chunks: Vec<usize> = (100..105).collect();
+    let losses: Vec<f32> = vec![0.1, 0.3, 0.5, 0.9];
+
+    // We should parameterize these as well
+    let robust_c = 0.2;
+    let robust_delta = 0.05;
+
+    for (size, chunk, loss) in izip!(&sizes, &chunks, &losses) {
```
Comment (JayKickliter) [inline benches/decoder_benchmark.rs]: This is certainly more clean than native Iter zipping, but I'm not sure it's worth bringing in a whole dependency for it.

Comment (vihu) [reply benches/decoder_benchmark.rs]: I thought a `dev-dep` should be fine...

Comment (JayKickliter) [reply benches/decoder_benchmark.rs]: Yeah, that's better than blakent dep. I guess it depends on how long it takes to fetch and build itertools (I don't know the answer to that).

File: benches/decoder_benchmark.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                    *chunk,
+                    *loss,
+                    robust_c,
+                    None,
+                    robust_delta,
+                    EncoderType::Systematic,
+                )
+            })
+        });
+        // group.bench_with_input(BenchmarkId::new("RobustConst", size), size,
```
Comment (JayKickliter) [inline benches/decoder_benchmark.rs]: Commented code will atrophy. Delete?

File: benches/decoder_benchmark.rs (hunk 1 above)
Comment (vihu) [reply benches/decoder_benchmark.rs]: It's actually wrong too, it should be `product!` ideally

File: benches/decoder_benchmark.rs (hunk 2 above)
Comment (vihu) [reply benches/decoder_benchmark.rs]: yea

**Review by JayKickliter — CHANGES_REQUESTED** [review]
From what I can see, there isn't much of a difference between RobustEncoder and IdealEncoder other than what solition they use. 

Considering that you added `trait Solition`, why just have one Encoder that takes a solition as an argument? If you're worried about `<T>` pollution, you could always make it a `Box<&dyn Solition>` at the cost of little performance.

**Comment by vihu** [comment]
> From what I can see, there isn't much of a difference between RobustEncoder and IdealEncoder other than what solition they use.
> 
> Considering that you added `trait Solition`, why just have one Encoder that takes a solition as an argument? If you're worried about `<T>` pollution, you could always make it a `Box<&dyn Solition>` at the cost of little performance.

I actually added an `Encoder` trait in https://github.com/vihu/fountain/pull/11. An encoder is just supposed to implement `next`. The reason I kept them separate (for now) was due to the `c` and `delta` extra params that the `RobustSoliton` takes in. But yeah, I was thinking about having just one Encoder as well which takes in all the required params to construct either RobustSoliton or IdealSoliton depending on the type of Soliton. I can do that in a separate PR, this is mostly just to get it there.

**Comment by JayKickliter** [comment]
> having just one Encoder as well which takes in all the required params to construct either RobustSoliton or IdealSoliton depending on the type of Soliton

Could also just have different constructors:
```
Encoder::robust(params..) -> Encoder
Encoder::ideal(params..) -> Encoder
```
The constructors would create the appotiate `Box<dyn Solition>`, and it as is one single type. Could also store the solition as an enum if derefing the solition is too expensive.

**Comment by vihu** [comment]
> > having just one Encoder as well which takes in all the required params to construct either RobustSoliton or IdealSoliton depending on the type of Soliton
> 
> Could also just have different constructors:
> 
> ```
> Encoder::robust(params..) -> Encoder
> Encoder::ideal(params..) -> Encoder
> ```
> 
> The constructors would create the appotiate `Box<dyn Solition>`, and it as is one single type. Could also store the solition as an enum if derefing the solition is too expensive.

Oh yea, that does seem much cleaner tbh.

**Review by JayKickliter — APPROVED** [review]

**Pull request merged**